### ISSUE TITLE
Enhance the kubeedge upgrading function

### ISSUE THREAD

**Issue by ZhengXinwei-F** [kubeedge/kubeedge#5094]
The existing kubeedge upgrading function should be enhanced.

- [ ] Designing a job structure is necessary. It is expected to help with the execution of jobs like node upgrade, rollback, backup, configuration update, and docker pre pull.
- [ ] Node rollback should be a separate feature.
- [ ] Node backup should be treated as a separate function that can be managed through cloudcore.  The scheduled backup function will be supported later.

**require discussion**

- [ ] Configuration will be lost during the node upgrade procedure. (We can design a CRD for configuration storage and configuration updates.)
- [ ] FSM is required and can control actions, events and status during the upgrade process.
```
    actions：upgrade、backup、rollback
    status: init -> backing up -> backup successful -> upgrading -> upgrade successful
                                                                 -> upgrade failed -> rolling back -> rollback successful
                                                                                                   -> rollback failed
                               -> backup failed
    events：init、success、failure、timeout
    FSM：

    (currentStatus/action/event: nextStatus)
    empty/upgrade/init: init
    init/backup/init: backing up
    backing up/backup/success: backup successful
    backing up/backup/failure: backup failed
    backup successful/upgrade/init: upgrading
    upgrading/upgrade/success: upgrade successful
    upgrading/upgrade/failure: upgrade failed
    upgrade failed/rollback/init: rolling back
    rolling back/rollback/success: rollback successful
    rolling back/rollback/failure: rollback failed
```

**other features to mention**
- [ ] configuration update
- [ ] docker pre pull

**Comment by WillardHu** [comment]
The definition of `NodeUpgradeJob.Spec` is basically satisfied, and we can optimize the status definition.

**Comment by ZhengXinwei-F** [comment]
> The PR #5085 resolves the upgrade node process issue, but it does not resolve the essential issue, and we need to either reopen this issue or create a new issue to trace. @ZhengXinwei-F

done.

I noticed that the current upgrading procedure does not include a check on the status of edgecore. Perhaps we can give a healthz api to let keadm know the status of the edgecore.

@WillardHu @wlq1212

**Comment by ZhengXinwei-F** [comment]
Furthermore, edgecore is the only communication link with the cloud. When an upgrade fails, the upgrade status is not normally reported, and the upgrade stage and state may differ between edge and cloud.  I think is the main issue at this point.

I haven't come up with a elegant approach yet, so you can actively contribute your ideas.

**Comment by ZhengXinwei-F** [comment]
Cloudcore's current version cannot be upgraded. We should provide keadm with the ability to upgrade cloudcore.

**Comment by ZhengXinwei-F** [comment]
In the current version, the keadm upgrading process is an orphan process. This strategy is insufficiently dependable, and unexpected disruptions will cause the edgecore update to fail.

**Comment by ZhengXinwei-F** [comment]
Is it required for us to employ multiple CRDs and controllers to accomplish various task activities in order to reduce user learning costs?

Like：
![node-task-in-one-controller](https://github.com/kubeedge/kubeedge/assets/77471374/3400748e-7b9d-4339-834a-e17efa0b5071)
![node-tasks-several-controllers](https://github.com/kubeedge/kubeedge/assets/77471374/29bbc500-5de4-44d8-8021-385c9bca7a04)

**Comment by ZhengXinwei-F** [comment]
Is it necessary to include OTA capabilities? Allow node administrators to decide whether or not to upgrade.

**Comment by WillardHu** [comment]
> Is it required for us to employ multiple CRDs and controllers to accomplish various task activities in order to reduce user learning costs?

![image](https://github.com/kubeedge/kubeedge/assets/4033081/5092c4f2-06be-4b47-960b-aff2ef97e4ba)

Perhaps we can use a workflow CRD to replace the logic of the `tasks message worker` module (WorkflowController?) and make the task CR reusable. @ZhengXinwei-F

**Comment by WillardHu** [comment]
When cloud init and edge join, we can use a manifest file to record the command args for easy execute the upgrade and reset.

**Comment by ZhengXinwei-F** [comment]
> > Is it required for us to employ multiple CRDs and controllers to accomplish various task activities in order to reduce user learning costs?
> 
> ![image](https://user-images.githubusercontent.com/4033081/279629878-5092c4f2-06be-4b47-960b-aff2ef97e4ba.png)
> 
> Perhaps we can use a workflow CRD to replace the logic of the `tasks message worker` module (WorkflowController?) and make the task CR reusable. The workflow CRD can also control the order in which tasks are executed at a edge node. @ZhengXinwei-F

Yes, I understand that your workflow comprises a number of different controllers, and we must consider task scalability in the design. A workflow with too many aggregated functions may make adding more tasks difficult.

**Comment by ZhengXinwei-F** [comment]
> When we execute the cloud init and edge join, we can use a manifest file to record the command args for easy execute the upgrade and reset.

Do you mean local recording on the edge node? If edgecore configuration is saved only locally, users will be unable to query and edit edgecore configuration from the cloud.

**Comment by WillardHu** [comment]
> > When we execute the cloud init and edge join, we can use a manifest file to record the command args for easy execute the upgrade and reset.
> 
> Do you mean local recording on the edge node? If edgecore configuration is saved only locally, users will be unable to query and edit edgecore configuration from the cloud.

![image](https://github.com/kubeedge/kubeedge/assets/4033081/dc6151b5-baef-4abd-8207-9047eae24c63)

Yes, it records some command args , which are basically related to the host environment. And we need to consider the upgrades and resets have multiple trigger scenarios.
During the upgrade, you can overwrite the recorded value by delivering new args. But maybe the clould doesn't care so much about these command args?

**Comment by ZhengXinwei-F** [comment]
> Yes, it records some command args, which are basically related to the host environment, and we need to consider the upgrades and resets have multiple trigger scenarios.
> During the upgrade, you can overwrite the recorded value by delivering new args. But maybe the clould doesn't care so much about these command args?

This causes an issue in that the user must access the node to find out about its configuration, which is inconvenient for the user's next setting.
I think the edgecore configurations can be saved in etcd as CR and declared automatically by edgecore.

**Comment by WillardHu** [comment]
"Construction of cloudcore upgrading capability" tracked by this issue https://github.com/kubeedge/kubeedge/issues/5131

**Comment by WillardHu** [comment]
> > Yes, it records some command args, which are basically related to the host environment, and we need to consider the upgrades and resets have multiple trigger scenarios.
> > During the upgrade, you can overwrite the recorded value by delivering new args. But maybe the clould doesn't care so much about these command args?
> 
> This causes an issue in that the user must access the node to find out about its configuration, which is inconvenient for the user's next setting. I think the edgecore configurations can be saved in etcd as CR and declared automatically by edgecore.

![image](https://github.com/kubeedge/kubeedge/assets/4033081/4228a956-6551-45a1-8c6b-de4411a6e944)

Okey, meybe we can set the record to node's annotations, so that we can meet all our requirements.

**Comment by ZhengXinwei-F** [comment]
> "Construction of cloudcore upgrading capability" tracked by this issue #5131

👍

**Comment by ZhengXinwei-F** [comment]
> Okey, meybe we can set the record to node's annotations, so that we can meet all our requirements.

This is good advise, and I will incorporate it into the design.

**Comment by ZhengXinwei-F** [comment]
Should the task CRD be designed to be cluster or namespace scope?
Some users lack cluster scope access and hence cannot create tasks.

**Comment by WillardHu** [comment]
> Should the task CRD be designed to be cluster or namespace scope? Some users lack cluster scope access and hence cannot create tasks.

I think the person who upgrades the node should have similar permissions to the person who joins the node. The node belongs to the cluster resources, and the scope of CRD should be defined as the cluster.

The authz of the k8s is used to define functional permissions(Use ClusterRole to define whether pod can access CRs read and write). Nodes and namespaces belong to data permissions, and on our platform, data permissions are handled using application layer.
